plugins/user.py

```diff
@@ -0,0 +1,220 @@
+        if hasattr(settings, "ADMIN_USERS"):
+            for admin_user in settings.ADMIN_USERS.split(','):
+                user_permissions[admin_user] = ['admin', 'grant']
+                storage.say("giving {} admin permissions".format(admin_user))
+            storage.save("user_permissions", user_permissions)
+        else:
+            msg = "No admin users are defined in redis or environment"
+            storage.say(msg)
+    if username in user_permissions:
+        if permission in user_permissions[username]:
+            return 1
+        else:
+            return 0
+    else:
+        return 0
+
+def verify_twofactor(storage, username, token):
+    user_twofactor = storage.load("user_twofactor", {})
+    if username in user_twofactor:
+        totp = pyotp.TOTP(user_twofactor[username])
+        if totp.verify(token):
+            return 1
+        else:
+            return 0
+    else:
+        return 0
+
+class UserPlugin(WillPlugin):
+    def __init__(self):
+        if not hasattr(settings, "TWOFACTOR_ISSUER"):
+            msg = "Error: TWOFACTOR_ISSUER not defined in the environment"
+            self.say(msg)
+        self.TWOFACTOR_ISSUER = settings.TWOFACTOR_ISSUER
+        if not hasattr(settings, "TWOFACTOR_PRINCIPLE"):
+            msg = "Error: TWOFACTOR_PRINCIPLE not defined in the environment"
+            self.say(msg)
+        self.TWOFACTOR_PRINCIPLE = settings.TWOFACTOR_PRINCIPLE
+        if not hasattr(settings, "TWOFACTOR_S3_BUCKET"):
+            msg = "Error: TWOFACTOR_S3_BUCKET not defined in the environment"
+            self.say(msg)
+        self.TWOFACTOR_S3_BUCKET = settings.TWOFACTOR_S3_BUCKET
+        if not hasattr(settings, "TWOFACTOR_S3_PROFILE"):
+            msg = "Error: TWOFACTOR_S3_PROFILE not defined in the environment"
+            self._say_error(msg)
+        self.TWOFACTOR_S3_PROFILE = settings.TWOFACTOR_S3_PROFILE
+
+
+            
+
+
+    def generate_and_upload_QR(self,secret,username):
+        try:
+            principle = self.TWOFACTOR_PRINCIPLE
+            issuer = self.TWOFACTOR_ISSUER
+            s3_twofactor_bucket = self.TWOFACTOR_S3_BUCKET
+            s3_twofactor_profile = self.TWOFACTOR_S3_PROFILE
+        except:
+            return "Settings for S3 hosted QR codes not configured properly in environment"
+
+        url = 'otpauth://totp/{issuer}:{principle}?secret={secret}&issuer={issuer}'.format(issuer=issuer, principle=principle, secret=secret)
+        img = qrcode.make(url)
+        conn = boto.connect_s3(profile_name=s3_twofactor_profile)
+        bucket = conn.get_bucket(s3_twofactor_bucket)
+        key = Key(bucket)
+        key.key = '{}-qr.png'.format(username)
+        image_buffer = cStringIO.StringIO()
+        img.save(image_buffer, 'PNG')
+        key.set_contents_from_string(image_buffer.getvalue())
+        url = key.generate_url(60,query_auth=True, force_http=True)
+        return url
+
+    @respond_to("^twofactor verify (?P<token>\w*)")
```

Comment:
Using `\w*` rather than `\w+` here could result in blank tokens getting matched.